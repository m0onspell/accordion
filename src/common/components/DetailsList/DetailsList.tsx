import React, { FC, memo } from 'react';
import styles from './DetailsList.module.css';
import startCase from 'lodash.startcase';

export interface DetailsListProps {
  data: Record<string, any>;
  renderFields?: string[];
}

const isNotLastItem = (index: number, fields: string[]) => index !== fields.length - 1;

const DetailsList: FC<DetailsListProps> = memo(({ data, renderFields = [] }) => {
  const fields = renderFields.length ? renderFields : Object.keys(data);
  return (
    <div className={styles.container}>
      {fields.map((field, index) => (
        <React.Fragment key={field}>
          <div className={styles.detailGroup}>
            <div>{startCase(field)}</div>
            <div>{data[field]}</div>
          </div>
          {isNotLastItem(index, fields) && <hr />}
        </React.Fragment>
      ))}
    </div>
  );
});

export { DetailsList };
