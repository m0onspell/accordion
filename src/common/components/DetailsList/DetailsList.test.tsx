import React from 'react';
import renderer from 'react-test-renderer';
import { DetailsList, DetailsListProps } from './DetailsList';

const makeProps = (props: Partial<DetailsListProps> = {}): DetailsListProps => ({
  data: {
    firstName: 'John',
    lastName: 'Doe',
    email: 'admin@mail.to',
  },
  ...props,
});

const setup = (props?: Partial<DetailsListProps>) =>
  renderer.create(<DetailsList {...makeProps(props)} />).toJSON();

test('renders correctly', () => {
  expect(setup()).toMatchSnapshot();
});

test('renders subset of fields', () => {
  expect(setup({ renderFields: ['email', 'lastName'] })).toMatchSnapshot();
});
