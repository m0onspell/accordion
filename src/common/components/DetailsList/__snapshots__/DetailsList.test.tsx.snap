// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="detailGroup"
  >
    <div>
      First Name
    </div>
    <div>
      John
    </div>
  </div>
  <hr />
  <div
    className="detailGroup"
  >
    <div>
      Last Name
    </div>
    <div>
      Doe
    </div>
  </div>
  <hr />
  <div
    className="detailGroup"
  >
    <div>
      Email
    </div>
    <div>
      admin@mail.to
    </div>
  </div>
</div>
`;

exports[`renders subset of fields 1`] = `
<div
  className="container"
>
  <div
    className="detailGroup"
  >
    <div>
      Email
    </div>
    <div>
      admin@mail.to
    </div>
  </div>
  <hr />
  <div
    className="detailGroup"
  >
    <div>
      Last Name
    </div>
    <div>
      Doe
    </div>
  </div>
</div>
`;
